src/System.Runtime/src/System/HashCode.cs

```diff
@@ -0,0 +1,406 @@
+            unchecked
+            {
+                hash += input * Prime2;
+                hash = Rol(hash, 13);
+                hash *= Prime1;
+                return hash;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint QueueRound(uint hash, uint queuedValue)
+        {
+            unchecked
+            {
+                hash += queuedValue * Prime3;
+                return Rol(hash, 17) * Prime4;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint MixState(uint v1, uint v2, uint v3, uint v4)
+        {
+            return Rol(v1, 1) + Rol(v2, 7) + Rol(v3, 12) + Rol(v4, 18);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint MixEmptyState()
+        {
+            return _seed + Prime5;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint MixFinal(uint hash)
+        {
+            unchecked
+            {
+                hash ^= hash >> 15;
+                hash *= Prime2;
+                hash ^= hash >> 13;
+                hash *= Prime3;
+                hash ^= hash >> 16;
+                return hash;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Add<T>(T value)
+        {
+            Add(value?.GetHashCode() ?? 0);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Add<T>(T value, IEqualityComparer<T> comparer)
+        {
+            if (comparer is null)
+                throw new ArgumentNullException(nameof(comparer));
+            Add(comparer.GetHashCode(value));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void Add(int value)
+        {
+            unchecked
+            {
+                var val = (uint)value;
+                var position = _length & 0x3;
+
+                switch (position)
+                {
+                    case 0:
+                        _queue1 = val;
+                        break;
+                    case 1:
+                        _queue2 = val;
+                        break;
+                    case 2:
+                        _queue3 = val;
+                        break;
+                    default:
+
+                        var sentinel =
+                            _v1 | _v2 | _v3 | _v4 | // any of accumulators set?
+                            (_length ^ position); // length greater than 3?
+
+                        if (sentinel == 0)
+                            Initialize(out _v1, out _v2, out _v3, out _v4);
+
+                        _v1 = Round(_v1, _queue1);
+                        _v2 = Round(_v2, _queue2);
+                        _v3 = Round(_v3, _queue3);
+                        _v4 = Round(_v4, val);
+
+                        break;
+                }
+
+                _length++;
```

Comment:
I missed the earlier comment about this overflowing. Could you just add an overflow check here and throw rather than silently resetting the state?